packages/payment-experience-api/src/api/paytrailOnlineRefundPaymentSuccessController.ts

```diff
@@ -0,0 +1,170 @@
+import type { Request, Response } from 'express'
+import { URL } from 'url'
+import { getOrderAdmin, getRefundAdmin } from '@verkkokauppa/order-backend'
+import {
+  cancelPaymentAdmin,
+  checkPaytrailRefundCallbackUrl,
+  getPaymentsForOrderAdmin,
+  Order,
+  PaymentStatus,
+} from '@verkkokauppa/payment-backend'
+import {
+  createUserRedirectUrl,
+  isAuthorized,
+  isCardRenewal,
+} from '../lib/paymentReturnService'
+import { parseUuidFromPaytrailRedirectCheckoutStamp } from '../lib/paytrail'
+import { parseMerchantIdFromFirstOrderItem } from '@verkkokauppa/configuration-backend'
+
+export class PaytrailOnlineRefundPaymentSuccessController extends AbstractController {
+  protected readonly requestSchema = null
+
+  protected async implementation(
+    request: Request,
+    result: Response
+  ): Promise<any> {
+    const { query } = request
+    // Validates that base redirect url is set
+    PaytrailOnlineRefundPaymentSuccessController.checkAndCreateRedirectUrl()
+    const refundId = parseUuidFromPaytrailRedirectCheckoutStamp({ query })
+
+    if (!refundId) {
+      logger.error(
+        'No refundId specified redirect to paytrail general failure url'
+      )
+      return result.redirect(
+        302,
+        PaytrailOnlineRefundPaymentSuccessController.getFailureRedirectUrl()
+      )
+    }
+
+    const refund = await getRefundAdmin({ refundId })
+    const orderId = refund.refund.orderId
+    const order = await getOrderAdmin({ orderId })
+    const merchantId = parseMerchantIdFromFirstOrderItem(order)
+
+    if (!merchantId) {
+      logger.error('Paytrail: No merchantId found from order')
+      throw new ExperienceError({
+        code: 'merchant-id-not-found',
+        message: 'No merchantId found from order.',
+        responseStatus: StatusCode.NotFound,
+        logLevel: 'info',
+      })
+    }
+
+    try {
+      const paytrailStatus = await checkPaytrailRefundCallbackUrl({
+        params: query,
+        merchantId: merchantId,
+      })
+      logger.debug(
+        `PaytrailStatus for refund ${refundId}: ${JSON.stringify(
+          paytrailStatus
+        )}`
+      )
+      if (!paytrailStatus.valid) {
+        logger.debug(
+          `PaytrailStatus is not valid for refund ${refundId}, redirect to failure url`
+        )
+        return result.redirect(
+          302,
+          PaytrailOnlineRefundPaymentSuccessController.getFailureRedirectUrl()
+        )
+      }
+
+      const redirectUrl = await createUserRedirectUrl({
+        order,
+        paymentReturnStatus: paytrailStatus,
+        redirectPaymentUrlBase: PaytrailOnlineRefundPaymentSuccessController.getRedirectUrl(),
+      })
+      // TODO: Make changes that sends email from refund
+      // Function contains internal checks when to send receipt.
+      // await sendReceiptToCustomer(paytrailStatus, orderId, order)
+
+      // Only cancel authorized card renewals.
+      if (isAuthorized(paytrailStatus) && isCardRenewal(paytrailStatus)) {
```

Comment:
You can delete this whole code block, I was wrong before when I said to leave this.